README.md
# BCScanner

![version tag](https://img.shields.io/github/tag/michaelochs/bcscanner.svg?style=flat-square)
![release](https://img.shields.io/github/release/michaelochs/bcscanner.svg?style=flat-square)
![cocoapods version](https://img.shields.io/cocoapods/v/BCScanner.svg?style=flat-square)
![open issues](https://img.shields.io/github/issues/michaelochs/bcscanner.svg?style=flat-square)

![apache 2.0 license](https://img.shields.io/cocoapods/l/BCScanner.svg?style=flat-square)
![iOS platform](https://img.shields.io/cocoapods/p/BCScanner.svg?style=flat-square)

`BCScanner` is a barcode and qr code scanner library that makes use of the built in `AVFoundation` code scanning capabilities. It lets you easily integrate a barcode or qr code scanner into your application without dealing with `AVFoundation`.


## Requirements

`BCScanner` is bild for the use with iOS7+. If you still do support iOS6, make sure to check the `+[BCScanner scannerAvailable]` method before making any other calls to `BCScanner`. `BCScanner` is not compatible with iOS6 because of the internal use of new `AVFoundation` API.
`BCScanner` also requires you to use ARC.


## Dependencies

- UIKit
- AVFoundation
- QuartzCore


## Install

### cocoapods

The easiest way to install this module is to add the library through cocoapods:

    pod 'BCScanner'

If you want to specify a certain version of `BCScanner` please note that this project uses [semantic versioning](http://semver.org). This means it is save to upgrade to every minor and patch release as this is fully backwards compatible. If you want to provide a version in your podfile, use this line:

    pod 'BCScanner', '~> 0.1'


### static library

If you, for some reason, do not want to directly integrate the project from the git repository, there is a [binary release in the release section on github](https://github.com/michaelochs/BCScanner/releases) for every version. This bundle includes a pre compiled library as well as a local `podspecs` file you can use to integrate the library locally.


### other integration

Currently, other types of integrations are not supported officially, however they are possible by simply using the library or the repository and adding `AVFoundation` as a dependency. Please also remember to add the `-ObjC` flag to your 'other linker flags' build settings.

See `BCScannerExample` target for a reference, if you have problems, linking the library.


## Example

To build the example project run `pod install` inside the `Example` folder and open up the `BCScanner.xcworkspace` file. Make sure you are using the workspace and not the project itself if you are experiencing issues.


## Contribution

If you are experiencing issues or have any question, please create an issue in the [issues section on github](https://github.com/michaelochs/BCScanner/issues). You can also reach me on twitter [@_mochs](https://twitter.com/_mochs) or check my blog at [ios-coding.com](http://ios-coding.com) for any how-to related questions.

Also feel free to open pull requests to improve the project. Thanks to everyone who has already done so in the past! For a list of contributors, please see the [CREDITS.md](https://github.com/michaelochs/BCScanner/blob/develop/CREDITS.md) file.


## Licence

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

CHANGELOG.md
# Changelog

The versioning in this project is based on [Semantic Versioning](http://semver.org).

## v0.1.1
- Fix issues with the scanner area property
- Reenables iOS7 support
- Removes warnings

## v0.1.0
- First official release, containing support for the following code types:
	- aztec
	- code 39
	- code 39 mod 43
	- code 93
	- code 128
	- data matrix (iOS8+)
	- EAN 8
	- EAN 13
	- Interleaved 2 of 5 (iOS8+)
	- ITF14 (iOS8+)
	- PDF 417
	- QR
	- UPC-E
	
- Support for customizing the appearance
- Convenience API for fast, single code scans

BCScanner/BCScannerViewController.h
//
//  BCScannerViewController.h
//  BCScannerViewController
//
//	Copyright 2013 bitecode, Michael Ochs
//
//	Licensed under the Apache License, Version 2.0 (the "License");
//	you may not use this file except in compliance with the License.
//	You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
//	Unless required by applicable law or agreed to in writing, software
//	distributed under the License is distributed on an "AS IS" BASIS,
//	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//	See the License for the specific language governing permissions and
//	limitations under the License.
//

#import <UIKit/UIKit.h>

@protocol BCScannerViewControllerDelegate;


#pragma mark - code types

// iOS7+
FOUNDATION_EXTERN NSString *const BCScannerUPCECode; /// The UPC-E code type.
FOUNDATION_EXTERN NSString *const BCScannerCode39Code; /// The code 39 code type.
FOUNDATION_EXTERN NSString *const BCScannerCode39Mod43Code; /// The code 39 code type with mod 43 support.
FOUNDATION_EXTERN NSString *const BCScannerEAN13Code; /// The EAN 13 code type.
FOUNDATION_EXTERN NSString *const BCScannerEAN8Code; /// The EAN 8 code type.
FOUNDATION_EXTERN NSString *const BCScannerCode93Code; /// The code 93 code type.
FOUNDATION_EXTERN NSString *const BCScannerCode128Code; /// The code 128 code type.
FOUNDATION_EXTERN NSString *const BCScannerPDF417Code; /// The PDF 417 code type.
FOUNDATION_EXTERN NSString *const BCScannerQRCode; /// The QR code type.
FOUNDATION_EXTERN NSString *const BCScannerAztecCode; /// The aztec code type.

// iOS8+
FOUNDATION_EXTERN NSString *const BCScannerI25Code; /// The Interleaved 2 of 5 code type. This type does nothing on iOS7, as it is only supported on iOS8 and later
FOUNDATION_EXTERN NSString *const BCScannerITF14Code; /// The ITF14 code type. This type does nothing on iOS7, as it is only supported on iOS8 and later
FOUNDATION_EXTERN NSString *const BCScannerDataMatrixCode; /// The data matrix code type. This type does nothing on iOS7, as it is only supported on iOS8 and later



#pragma mark - controller

/**
 BCScannerViewController is a view controller that wrapps the scanning
 capabilities of iOS7 into a simple to use drop-in view controller for easy
 integration into your app.
 
 It is build for the purpose of scanning a single or multiple codes in a
 deticated screen and encapsules all the camera handling and metadata gathering.
 
 The view controller can be presented or pushed onto a navigation stack. It is
 up to you to present and dismiss the view controller whenever you need it.
*/
@interface BCScannerViewController : UIViewController

/**
 This method lets you check for the availability of the native scanner
 functionality provided by AVFoundation.

 @note Do not instanciate this class if this method returns NO!

 @return YES if native scanning is available, NO otherwise.
 */
+ (BOOL)scannerAvailable;

/**
 This is the delegate of the scanner. The delegate will get called to notify you
 about codes that were found in the field of view of the camera.
 */
@property (nonatomic, weak, readwrite) id<BCScannerViewControllerDelegate> delegate;

/**
 This is an array of the codes the scanner should look for. The more code types
 you specify, the longer the image analysis will take!

 @see BCScannerCode constants
 */
@property (nonatomic, strong, readwrite) NSArray *codeTypes;

/**
 This is the gesture recognizer that is used to let the user focus and expose to
 a specific point in the field of view. You can access this property when you
 want to more precisely control what taps should be recognized.
 */
@property (nonatomic, weak, readonly) UITapGestureRecognizer *focusAndExposeGestureRecognizer;

/**
 Defines the rect of the view controller's view that describes the active area
 that is used by `AVFoundation` to scan for barcodes.
 
 The default value is `CGRectZero` which turns this feature off and always uses
 the whole camera input as a capturing area.

 @note all codes that are completely visible by the active camera and intersect
       with the specified area are scanned. The code does not need to be
       completely inside the scanner area.
 */
@property (nonatomic, assign, readwrite) CGRect scannerArea;



#pragma mark - torch

/**
 Defines the video preview torch mode.
 */
@property (nonatomic, assign, readwrite, getter = isTorchEnabled) BOOL torchEnabled;

/**
 Hides or displays the torch button in the navigation bar.
 
 The default value is YES.
 */
@property (nonatomic, assign, readwrite, getter = isTorchButtonEnabled) BOOL torchButtonEnabled;

/**
 Indicates if the torch mode is available. Varies depending on the device.
 */
@property (nonatomic, assign, readonly, getter = isTorchModeAvailable) BOOL torchModeAvailable;

@end



#pragma mark - delegate

@protocol BCScannerViewControllerDelegate <NSObject>

@optional
/**
 This method is called whenever a new code enters the field of view.

 @param	scanner	The scanner that is calling this delegate
 @param	codes	A list of all the codes that entered the FOV in this interval

 @note	If you do a simple scan for the first code you find, you can get the
        code from this method and close the scanner afterwards.
 */
- (void)scanner:(BCScannerViewController *)scanner codesDidEnterFOV:(NSSet *)codes;

//- (void)scanner:(BCScannerViewController *)scanner codesDidUpdate:(NSSet *)codes;

/**
 This method is called whenever an existing code leaves the field of view.

 @param	scanner	The scanner that is calling this delegate
 @param	codes	A list of all the codes that left the FOV in this interval
 */
- (void)scanner:(BCScannerViewController *)scanner codesDidLeaveFOV:(NSSet *)codes;

/**
 This method lets you specify an image that is shown as an overlay to give the
 user some feedback about how to hold the camera.

 This method is called when the scanner configures its interface.

 @param scanner	The scanner that is calling this delegate
 
 @return The image you want to be used as HUD
 */
- (UIImage *)scannerHUDImage:(BCScannerViewController *)scanner;

@end
